mlir/include/mlir/Dialect/Rock/IR/GetRockInfo.h

```diff
@@ -0,0 +1,104 @@
+//===- GetRockInfo.h - functions used to calculate information about Rock ops
+//---------------===//
+//
+// Part of the MLIR Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#ifndef MLIR_DIALECT_ROCK_IR_GETROCKINFO_H
+#define MLIR_DIALECT_ROCK_IR_GETROCKINFO_H
+
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/Value.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "mlir/Support/LLVM.h"
+
+namespace mlir {
+class Operation;
+class Type;
+ 
+namespace rock {
+
+// Get the arch from the op
+FailureOr<StringAttr> getArch(Operation *op);
+
+// Get the arch from the op and error out if it cannot be found
+StringAttr getArchValue(Operation *op);
+
+// Get the num_cu from the op
+FailureOr<int64_t> getNumCU(Operation *op);
+
+// Get the num_cu from the op, and error out if it cannot be found
+int64_t getNumCUValue(Operation *op);
+
+// Get the features enabled for the specified op. These will be dependent on
+// the architecture being used, and the type of the op.
+rock::GemmFeatures getFeatures(Operation *op);
+
+// This function returns a boolean value if the underlying op has support for
+// an optional 'GemmFeatures' attribute
+bool opHasOptionalFeature(Operation *op);
+
+rock::GemmFeatures insersectGemmFeatures(rock::GemmFeatures a,
+                                         rock::GemmFeatures b);
+
+// This function returns the func or gpu.func of a given op
+inline Operation *getParentFuncOp(Operation *op) {
+  Operation *func;
+  if (isa<func::FuncOp, gpu::GPUFuncOp>(op)) {
+    func = op;
+  } else {
+    func = op->getParentOfType<func::FuncOp>();
+    if (!func) {
+      func = op->getParentOfType<gpu::GPUFuncOp>();
+    }
+  }
+
+  return func;
+}
+
+// Helper function to get attributes from parents
+template <typename RetAttrType>
+FailureOr<RetAttrType> getAttrFromOpOrParents(
+    Operation *op, StringRef opAttr,
+    std::optional<StringRef> maybeDialectAttr = std::nullopt) {
+  StringRef dialectAttr = maybeDialectAttr.value_or(opAttr);
+  Operation *func = getParentFuncOp(op);
+  RetAttrType attr;
+  auto getAnyAttr = [&](ArrayRef<StringRef> attrNames, Operation *op) {
+    for (StringRef attrName : attrNames) {
+      if (!attr) {
+        attr = op->getAttrOfType<RetAttrType>(attrName);
+      } else {
+        return;
+      }
+    }
+  };
+  getAnyAttr({opAttr}, op);
+  if (!attr) {
+    getAnyAttr({opAttr, dialectAttr}, func);
+  }
```

Comment:
```suggestion
 getAnyAttr({opAttr, dialectAttr}, func);
```

wouldn't something like work ? `getAnyAttr` is doing early exist if it finds `opAttr`. if doesn't then it would go and try to find attribute from `dialectAttr`

Reply:
I don't think we can remove this because the first `getAnyAttr({opAttr}, op);` is checking for the attribute on the op, and the second call to `getAnyAttr` is checking on the function.

For additional context, this was an already existing method that was in lowerinUtils.cpp that I moved over to GetRockInfo.